### ISSUE TITLE
`codecs.encode` with `utf-*` encoding and errors returing `str` rejects surrogates blindly

### ISSUE THREAD

**Issue by litlighilit** [python/cpython#127305]
# Bug report

### Bug description:

For [`codecs.encode`](https://docs.python.org/3/library/codecs.html#codecs.encode),
with `utf-*` encoding, and a custom `errors` which returns `str`,
if you pass some characters that are not valid UTF characters (e.g. surrogates),
`UnicodeEncodeError` is just raised and there's not the expected (and documented) case
where the returned `str` is appended.

```python
import codecs

ERRORS_NAME = "returning non-ascii"

# something being not encod-able via `utf-*`
BAD_UTF = "\uD800"  # the first high surrogate character


def register_repl_error(repl: str):
    def error_handle(exc: UnicodeEncodeError) -> tuple[str, int]:
        return (repl, exc.end)

    codecs.register_error(ERRORS_NAME, error_handle)


def encode_surrogate(encoding: str, repl: str):
    register_repl_error(repl)
    max_enc_len = 9
    pre = f"codecs.encode({BAD_UTF!r}, {encoding=:{max_enc_len}}) "
    try:
        res = codecs.encode(BAD_UTF, encoding, ERRORS_NAME)
    except UnicodeEncodeError as err:
        reason = err.reason
        print(pre + f"raises with {reason=}")

    else:
        print(pre + f"returns {res}")


NON_ASCII = "龍"  # loong in Chinese

## utf-*

for i in ('8', '16', '32', '16-le', '16-be'):
    encode_surrogate("utf-" + i, NON_ASCII)

print('-'*3)


# The following is some non-utf* encoding, which works fine

## cjk

### zh
for enc in ("gbk", "big5"):
    encode_surrogate(enc, NON_ASCII)

### jp
for enc in ("Shift_JIS", "EUC-JP"):
    encode_surrogate(enc, NON_ASCII)

```

Output:

```txt
codecs.encode('\ud800', encoding=utf-8    ) raises with reason='surrogates not allowed'
codecs.encode('\ud800', encoding=utf-16   ) raises with reason='surrogates not allowed'
codecs.encode('\ud800', encoding=utf-32   ) raises with reason='surrogates not allowed'
codecs.encode('\ud800', encoding=utf-16-le) raises with reason='surrogates not allowed'
codecs.encode('\ud800', encoding=utf-16-be) raises with reason='surrogates not allowed'
---
codecs.encode('\ud800', encoding=gbk      ) returns b'\xfd\x88'
codecs.encode('\ud800', encoding=big5     ) returns b'\xc0s'
codecs.encode('\ud800', encoding=Shift_JIS) returns b'\x97\xb4'
codecs.encode('\ud800', encoding=EUC-JP   ) returns b'\xce\xb6'
```

### CPython versions tested on:

3.9, 3.11, 3.12, 3.13, 3.14

### Operating systems tested on:

Linux, Windows

### Linked PRs
* gh-139554

**Comment by litlighilit** [comment]
I'm going to fix it until I realize something serious to communicate: **how errors shall be handled within error_handle?***

One is simply re-call `encode` with 'strict' errors against the returned `str`, which, however, may lead to dead loop (recursion), because `encode` will still invoke error_handler against the same data of `str`

Or maybe using other `errors`?

In my option using `strict` errors with some ticks to avoid deadloop is the best.

**Comment by picnixz** [comment]
I'll have a look at it in a few days. I think error handler should not have errors but I need to verify. If the error handler errs, then the exception should be propagated.

**Comment by litlighilit** [comment]
> I'll have a look at it in a few days. I think error handler should not have errors but I need to verify. If the error handler errs, then the exception should be propagated.

Right, and it in fact does as expected, as edited above, my previous option was incorrect.

---

Then let's return to the original issue:

I've figured out where the mistaken code lies, and there're at least three places required to be fixed.

Yet these two days I'm too busy to focus on this patch, sorry in advance but I'll make it this weekday.

**Comment by StanFromIreland** [comment]
The current behaviour is actually tested:
https://github.com/python/cpython/blob/ff0cf0af10a775ea03dc2546b6a5506a708c2ab7/Lib/test/test_codeccallbacks.py#L815-L827

cc @serhiy-storchaka

**Comment by litlighilit** [comment]
> The current behaviour is actually tested:
> https://github.com/python/cpython/blob/ff0cf0af10a775ea03dc2546b6a5506a708c2ab7/Lib/test/test_codeccallbacks.py#L815-L827
> 
> cc @serhiy-storchaka 
> 
> 

If I did not misunderstand, the test is not the case in which the issue talks.

As what the errors handler returns is an valid UTF character ( `\U0001f40d`, i.e. '🐍')

**Comment by litlighilit** [comment]
> The current behaviour is actually tested:
> https://github.com/python/cpython/blob/ff0cf0af10a775ea03dc2546b6a5506a708c2ab7/Lib/test/test_codeccallbacks.py#L815-L827
> 
> cc @serhiy-storchaka 
> 
> 

The test is not the case in which the issue talks.

As in these tests, what the errors handler returns is a valid char ( `\U0001f40d`, i.e. '🐍') encodable via given encoding

**Comment by litlighilit** [comment]
I just realize the only issue is we haven't documented what errors will we use for the returned string.

What we have been doing is using 'strict'

May we just add a warning/note section to doc of [register_error](https://docs.python.org/3/library/codecs.html#codecs.register_error)

**Comment by serhiy-storchaka** [comment]
The original test was added in 9ab7dd4d5b616cd9486d3a73a5e8320f8c913a1a. I only significantly extended it in 18b07d773e09a2719e69aeaa925d5abb7ba0c068.

Why is the non-ASCII replacement an error? I guess just for code simplicity and because there wasn't a use case for this. In Python 2, 8-bit strings and Unicode strings are interchangeable if they contain only ASCII characters. That's why it made sense to support such Unicode strings.

If Asian multibyte codecs use recursive encoding for replacement string, then this can be done for other codecs as well. But there are pitfalls: UTF-16 and UTF-32 codecs without the byteorder qualifier produce a BOM -- but we don't want to insert a BOM in the middle of string.
